DotBot.py

```diff
@@ -176,226 +176,119 @@ def _setSpeed(self, speed):
+                if 90 < heading < 180:
+                    # bottom right quadrant
 
-                if q[i] < 0:
-                    return (None, None, None)
-                pass  # nothing to do
-            else:
-                t = q[i] / p[i]
-                if (p[i] < 0 and u1 < t):
-                    u1 = t
-                elif (p[i] > 0 and u2 > t):
-                    u2 = t
-
-        # if I get here, u1 and u2 should be set
-        assert u1 is not None
-        assert u2 is not None
-
-        # decide what to return
-        if (u1 >= 0 and u1 < u2 and u2 <= 1):
-
-            bump_x      = rx + u1 * deltax
-            bump_y      = ry + u1 * deltay
-            
-            timetobump  = u.distance((rx, ry), (bump_x, bump_y)) / self.currentSpeed
-            bump_ts     = self.tsMovementStart + timetobump
-            log.debug(f'Dotbot {self.dotBotId} obstacle bump at ({bump_x},{bump_y})')
-            assert bump_x >= 0 and bump_y >= 0
-            return (bump_x, bump_y, bump_ts)
+                    topX    = ((ymin - currentY) / slope) + currentX       # intersection with top obstacle border    (y=ymin)
+                    leftY   = (xmin - currentX)*slope + currentY           # intersection with left obstacle border   (x=xmin)
+
+                if 180 < heading < 270:
+                    # bottom left quadrant
+
+                    topX    = ((ymin - currentY) / slope) + currentX        # intersection with top obstacle border    (y=ymin)
+                    rightY  = (xmax - currentX)*slope + currentY            # intersection with right obstacle border  (x=xmax)
+
+                if 270 < heading < 360:
+                    # top left quadrant
+                    bottomX = ((ymax - currentY) / slope) + currentX       # intersection with bottom obstacle border (y=ymax)
+                    rightY  = (xmax - currentX)*slope + currentY           # intersection with right obstacle border  (x=xmax)
+
+                # check if intersection points are on obstacle
+                if topX and xmin <= round(topX, 10) <= xmax:
+                    intersectPoints += [(topX, ymin)]
+                if bottomX and xmin <= round(bottomX, 10) <= xmax:
+                    intersectPoints += [(bottomX, ymax)]
+                if leftY and ymin <= round(leftY, 10) <= ymax:
+                    intersectPoints += [(xmin, leftY)]
+                if rightY and ymin <= round(rightY, 10) <= ymax:
+                    intersectPoints += [(xmax, rightY)]
+
+        if intersectPoints:
+
+            # find closest intersection point to current position
+            distances      = [((x,y), u.distance((currentX, currentY), (x, y))) for (x, y) in intersectPoints]
+            distances      = sorted(distances, key=lambda e: e[1])
+            (bumpX, bumpY) = distances[0][0]
+
+            # find bump time
+            timetobump     = u.distance((currentX, currentY), (bumpX, bumpY)) / speed if speed != 0 else None
```

Comment:
you should `assert speed!=0` at the very top of the function to make sure you don't call this function when the DotBot isn't moving